### ISSUE TITLE
Speed issue in v0.9.1

### ISSUE THREAD

**Issue by licode** [lmfit/lmfit-py#259]
I did a test to compare new version and v0.8.3 as the following example.

```
def gaussian(x, amp, cen, wid):
    return amp * np.exp(-(x-cen)**2 /wid)

t1 = time.time()
gmode = None

for i in range(300):
    prename = 'g'+str(i)+'_'
    gmode1 = lmfit.Model(gaussian, prefix=prename)
    if gmode is None:
        gmode = gmode1
    else:
        gmode += gmode1
        
t2 = time.time()
p = gmode.make_params()
t3 = time.time()
copy.deepcopy(p)
t4 = time.time()

print(t2-t1)
print(t3-t2)
print(t4-t3)
```

result for v0.9.1 is
```
20.6462230682
1.95246195793
1.74573016167
```

while result for v0.8.3 is
```
0.400651931763
0.0140390396118
0.0666930675507
```

Any reason why the new version is slow in building model and creating parameters? I checked that lots of deepcopy are used in order to keep input parameters separated from output parameters. 

I wrote some complex model to do fitting with many gaussian functions, but right now it becomes extremely slow with the new version. Anyway we can turn this deepcopy off based on user's requirement? I can work on this if we care more about the speed issue.

**Comment by newville** [comment]
@licode   Hmm,  I'm not sure, but I'm also not sure it's only due to deepcopy, but I suspect (but have not confirmed) that this slowdown probably happened around #214.    

That is, Parameters() got more expensive to create, and each Model creates an OrderedSet() for its parameter names.   I think both of these are contributing.   But, deepcopy is definitely suspicious too!

I'm on travel at the moment, and just don't have time to look into this further.    Let us know what you figure out. 

Perhaps we need a longer release candidate time scale.  We're clearly missing important problems.

**Comment by tacaswell** [comment]
@licode You should use the `%prun` ipython magic to investigate what changed.

**Comment by tritemio** [comment]
@licode @tacaswell another very useful tool is `line_profiler`. Pip install it an it enables `%lprun` magic to have line-by-line performance stats.

**Comment by andyfaff** [comment]
I've been looking at overhead in Parameter._getval() recently. Testing
indicates that the total time required to obtain parameter values is a
significant fraction (~1/3) of the time it takes to calculate the objective
function.
It might be worth you looking there as well.

On 6 October 2015 at 11:51, Antonino Ingargiola <notifications@github.com>
wrote:

> @licode <https://github.com/licode> @tacaswell
> <https://github.com/tacaswell> another very useful tool is line_profiler.
> Pip install it an it enables %lprun magic to have line-by-line
> performance stats.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lmfit/lmfit-py/issues/259#issuecomment-145711973>.
>



-- 
_____________________________________
Dr. Andrew Nelson


_____________________________________

**Comment by ibressler** [comment]
Hi,
I'm new to lmfit and it looks very promising (thanks for that!) but I'm experiencing those performance issues as well.
Until now we used scipy.leastsq for basic background&scaling fit in each iteration of a Monte-Carlo procedure (http://mcsas.net). For an upcoming extension, I planned to use LM fit with box constraints and gave the recent version of lmfit 0.9.1 a try which increased its run time by ~6x and brought me here - so far the background.

After some profiling (cProfile) of our parent scope procedure calling lmfit >1e3 times (please see below for profiling output), it seems that providing the Parameter instances per tuple to lmfit.minimize() reduces the required time to ~50% (by avoiding deepcopy) in my case.
For example:
```
scaling = lmfit.Parameter('scaling', value = 1.0)
background = lmfit.Parameter('background', value = 0.1)
lmfit.minimize(residual, (scaling, background), args = (dataMeas, dataErr, dataCalc))
```

What is left are rather expensive calls to the Parameters `__init__` when creating the instance for the result object. The profiler pointed me to the `asteval.Interpreter.__init__()` method.
Therefore, even if I need box constraints only it creates and manages the whole (empty) meta data structure.

A possible approach might be to enable such expensive features only if they are requested via the interface. E.g. if there were no constraints (equations?) specified none of this needs to be handled ...

This is my profiling output for the parameters-by-tuple case:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.625    0.625   82.112   82.112 ~/code/mcsas/mcsas/mcsas.py:382(mcFit)
     6586    0.327    0.000   65.191    0.010 ~/code/mcsas/mcsas/backgroundscalingfit.py:99(fitLM)
     6585    0.073    0.000   63.232    0.010 ~/code/mcsas/lmfit/minimizer.py:706(minimize)
     6585    0.147    0.000   63.113    0.010 ~/code/mcsas/lmfit/minimizer.py:653(minimize)
     6585    1.555    0.000   62.894    0.010 ~/code/mcsas/lmfit/minimizer.py:501(leastsq)
     6585    0.488    0.000   39.481    0.006 ~/code/mcsas/lmfit/minimizer.py:309(prepare_fit)
     6585    0.204    0.000   34.316    0.005 ~/code/mcsas/lmfit/parameter.py:38(__init__)
     6585   18.193    0.003   33.110    0.005 ~/code/mcsas/lmfit/asteval.py:72(__init__)
     6585    0.615    0.000   17.145    0.003 /usr/lib/python2.7/dist-packages/scipy/optimize/minpack.py:255(leastsq)
    56185    1.834    0.000   10.976    0.000 ~/code/mcsas/lmfit/minimizer.py:240(__residual)
     6585    0.511    0.000   10.128    0.002 {scipy.optimize._minpack._lmdif}
    13167    1.089    0.000    9.449    0.001 ~/code/mcsas/mcsas/mcsas.py:818(calcModel)
   197975    1.210    0.000    7.186    0.000 ~/code/mcsas/lmfit/parameter.py:478(value)
    56185    0.369    0.000    6.583    0.000 ~/code/mcsas/mcsas/backgroundscalingfit.py:107(residual)
   197975    3.392    0.000    5.976    0.000 ~/code/mcsas/lmfit/parameter.py:438(_getval)
[...]
```

... and the one for providing parameters-by-Parameters-instance case:
```
   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.460    0.460   99.790   99.790 ~/code/mcsas/mcsas/mcsas.py:382(mcFit)
     5302    0.292    0.000   86.277    0.016 ~/code/mcsas/mcsas/backgroundscalingfit.py:99(fitLM)
     5301    0.058    0.000   57.376    0.011 ~/code/mcsas/lmfit/minimizer.py:707(minimize)
     5301    0.122    0.000   57.280    0.011 ~/code/mcsas/lmfit/minimizer.py:654(minimize)
     5301    1.183    0.000   57.109    0.011 ~/code/mcsas/lmfit/minimizer.py:502(leastsq)
    10603    0.270    0.000   50.556    0.005 ~/code/mcsas/lmfit/parameter.py:38(__init__)
    10603   27.178    0.003   48.964    0.005 ~/code/mcsas/lmfit/asteval.py:72(__init__)
     5301    0.246    0.000   40.713    0.008 ~/code/mcsas/lmfit/minimizer.py:310(prepare_fit)
79515/5301    0.706    0.000   39.179    0.007 /usr/lib/python2.7/copy.py:145(deepcopy)
     5301    8.886    0.002   38.953    0.007 ~/code/mcsas/lmfit/parameter.py:45(__deepcopy__)
     5301    0.461    0.000   11.975    0.002 /usr/lib/python2.7/dist-packages/scipy/optimize/minpack.py:255(leastsq)
    44485    1.343    0.000    8.028    0.000 ~/code/mcsas/lmfit/minimizer.py:241(__residual)
     5301    0.342    0.000    7.359    0.001 {scipy.optimize._minpack._lmdif}
[...]
```

... just some rough debugging last night.
Keep up the good work and best regards!

**Comment by andyfaff** [comment]
I've profiled `asteval` initiation [here](https://gist.github.com/andyfaff/3af1f722b0032edef6e2). Approximately 35% of the time is adding all the numpy 'symbols' to the `symtable`. This is done in a for loop and some speedup may be possible from a `dict` comprehension. About 6% is adding python symbols, for which some speed up may also be possible. Around 54% of the time is adding all those symbols into a `no_deepcopy` list.  Again, that is done in a `for` loop and a list comprehension might be faster.  @newville, it's not clear to me what the `no_deepcopy` list is for. Is it to speed up copying of the interpreter?

**Comment by andyfaff** [comment]
At the moment creating the `no_deepcopy` list in `asteval.__init__` takes up 54% of the time, 2980us, when creating the `Interpreter`. 

I've just [experimented](https://gist.github.com/andyfaff/f7aba2aeaddbdc67d7b6) (not tested to see if it works) with a list comprehension and it may be possible to reduce that to 222 us.

**Comment by newville** [comment]
@andyfaff  thanks for the profiling and testing!   This might be sort of challenging to get everything back to "right".    I'm pretty sure that OrderedSet() in Model() is part but definitely not all) of the problem @licode was seeing.   I think it's OK too simply make that a regular set(), and order the parameters later (perhaps only for printing).

**Comment by andyfaff** [comment]
I'm creating a test branch in my fork to make those list/dict
comprehensions in asteval.__init__. If it seems to speed stuff up I'll
submit a PR.

On 7 October 2015 at 08:38, Matt Newville <notifications@github.com> wrote:

> @andyfaff <https://github.com/andyfaff> thanks for the profiling and
> testing! This might be sort of challenging to get everything back to
> "right". I'm pretty sure that OrderedSet() in Model() is part but
> definitely not all) of the problem @licode <https://github.com/licode>
> was seeing. I think it's OK too simply make that a regular set(), and order
> the parameters later (perhaps only for printing).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lmfit/lmfit-py/issues/259#issuecomment-146009235>.
>



-- 
_____________________________________
Dr. Andrew Nelson


_____________________________________

**Comment by andyfaff** [comment]
Ok, so I've managed to get quite a lot done today.

* By using dict comprehensions it's possible to get a reasonable speedup on creating `asteval.Interpreter` instances.
* However, the real reason for a lot of these slowdowns was the code contained in `__deepcopy__`. To start with there was a sizeable chunk of time finding symbols in `_asteval.symtable` that were added by users, not during the original construction of the interpreter.  I've sped this up by replacing the for loops by a set difference. This slowdown was roughly constant with the number of added parameters.
* The real (huge) slowdown was due to addition of multiple parameters to the copied parameter object, `_pars`. Every time a new parameter is added to `_pars` the `update_constraints` method was triggered. The `update constraints` method runs through all the parameters sequentially in a for loop. Thus, the for loop runs once when the first parameter is added, 10 times when the 10th parameter is added, and so on. This is probably O(N**2) behaviour.  In reality the `update_constraints` loop only needs to run once, when all the parameters have been added. I've therefore added a flag, `_dont_update_constraints`, which is normally set to `False`, but is set to `True` when adding multiple parameters.  Hopefully this fix is acceptable.

I get the following times when reproducing @licode code:

    18.462864875793457
    0.1937391757965088
    0.026486873626708984

i.e. I've fixed `Parameters.__deepcopy__` and `Model.make_params`.  However, I'm yet to find a fix for `Model()` creation and appending. PR will come in the next few days

**Comment by licode** [comment]
@andyfaff , thanks for those updates. It looks promising. I didn't know dict comprehension works so efficiently. 

Inside Model, I would like to recommend checking all the places where deepcopy is used. It may be a good idea to also add a flag to turn deepcopy part on and off based on user's needs.

**Comment by andyfaff** [comment]
Please see #261 for a PR that addresses a lot of these problems.

**Comment by tacaswell** [comment]
@licode A flag the toggles deep-copy is a dangerous thing.  Either your state is immutable and you don't need to, or it is mutated by the process and you _must_ deep copy.

**Comment by andyfaff** [comment]
The flag I've added is called `_dont_update_constraints`. If this is set to
`True` then `update_constraints` doesn't do anything.  Whenever many
`Parameter` are added to a `Parameters` object (such as in `__deepcopy__`)
this is set to `True`. Once the Parameters have finished being added the
flag is set back to `False` and `update_constraints` is called.

On 7 October 2015 at 15:19, Thomas A Caswell <notifications@github.com>
wrote:

> @licode <https://github.com/licode> A flag the toggles deep-copy is a
> dangerous thing. Either your state is immutable and you don't need to, or
> it is mutated by the process and you *must* deep copy.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/lmfit/lmfit-py/issues/259#issuecomment-146073760>.
>



-- 
_____________________________________
Dr. Andrew Nelson


_____________________________________

**Comment by licode** [comment]
PR #261 looks good to me, except that we don't want to change the prefix name once model is created. Thanks for the work!

Probably next goal is try to optimize OrderedSet, especially self.add function there. It takes lots of time when we add models together. Any other option than OrderedSet for the basic data structure?

**Comment by newville** [comment]
I believe these issues have all been addressed in #261.   Thanks @andyfaff !
